### DISCUSSION TITLE
React Native App Hangs When Uploading File

### DISCUSSION THREAD

**Discussion by EverlyScott** [pocketbase/pocketbase#2002] [Q&A] [answered]
Hello, I am trying to write a React Native app which is using a PocketBase backend. I'm implementing a feature that allows users to change their profile picture, but I'm having an issue where the process hangs and uses a ton of resources (250% CPU usage & 4GB of RAM, which only increases over time). Here is my code which asks the user to choose a pfp from their photo library and uploads it:
```tsx
// Libraries in use:
import FormData from "form-data";
import * as ImagePicker from "expo-image-picker";
import { Buffer } from "buffer"; // internal node buffer library

  const handlePickPfp = async () => {
    setLoading(true);
    setError(false);
    try {
      await ImagePicker.getMediaLibraryPermissionsAsync(false);
      let result = await ImagePicker.launchImageLibraryAsync({
        mediaTypes: ImagePicker.MediaTypeOptions.Images,
        allowsEditing: true,
        aspect: [1, 1],
        quality: 1,
        base64: true,
        exif: false,
      });

      if (!result.canceled && result.assets[0].base64) {
        const data = new FormData();
        console.log(1);

        const buffer = Buffer.from(result.assets[0].base64, "base64");
        console.log(2);

        data.append("avatar", buffer);
        console.log(3);

        await pb.collection("users").update<IUser>(currentUser?.model?.id ?? "", data);
        console.log(4);
      }
      setLoading(false);
    } catch (err) {
      setLoading(false);
      setError(true);
      console.error(err);
    }
  };
  ```

The console outputs the following when the code is run:
```
 LOG  1
 LOG  2
 ```

but if I remove the last line which uploads it like this:
```tsx
      if (!result.canceled && result.assets[0].base64) {
        const data = new FormData();
        console.log(1);

        const buffer = Buffer.from(result.assets[0].base64, "base64");
        console.log(2);

        data.append("avatar", buffer);
        console.log(3);

        // await pb.collection("users").update<IUser>(currentUser?.model?.id ?? "", data);
        // console.log(4);
      }
```

I get the following output from the console:
```
 LOG  1
 LOG  2
 LOG  3
```
and the function completes successfully. I don't understand why removing the last line would make the code execute further when it apparently hasn't even run the code either time, and I don't understand why it wouldn't be working in the first place.

If I change the code like the following, everything works on the app side:
```diff
- const buffer = Buffer.from(result.assets[0].base64, "base64");
+ const buffer = Buffer.from(result.assets[0].base64, "base64").toString("ascii"); // or any format other than a buffer
```
 but I get a 413 error back (even though the size of the data sent is less than the 50mb limit):
(I'm not expecting anything to work on the PocketBase side when sending it via ascii, just experimenting to see if I can get the request through on the client, which it did)

**Comment by ganigeorgiev** [comment]
I don't have much experience with React Native or the Expo components so I'm not sure what could be the reason for the hang. It could be even something specific to the React Native `fetch` implementation.

One note to keep in mind is that the base64 conversion increases the size of the parsed file with ~30% and I don't recommend it for files larger than 10MB.

What you can try is, instead of creating a buffer from the encoded base64 file content, to disable the ImagePicker base64 conversion and to fetch its binary data as noted in https://github.com/pocketbase/pocketbase/discussions/1916#discussioncomment-5109710, aka:

```js
const response = await fetch(result.assets[0].uri);
const blob = await response.blob();
```

**Reply by ganigeorgiev** [reply]
Also one more thing - it is better to use the native `FormData` implementation (I believe it should be available in React Native with node17+) since the custom `form-data` dependency you load may not be compatible with the `fetch` that the SDK is using for sending the request.

**Reply by EverlyScott** [reply]
Doing this has got me further, but now when Pocketbase tries to upload the profile picture, it just sets `avatar` to an empty string.

Here's my code now:
```tsx
  const handlePickPfp = async () => {
    setLoading(true);
    setError(false);
    try {
      await ImagePicker.getMediaLibraryPermissionsAsync(false);
      let result = await ImagePicker.launchImageLibraryAsync({
        mediaTypes: ImagePicker.MediaTypeOptions.Images,
        allowsEditing: true,
        aspect: [1, 1],
        quality: 1,
        base64: false,
        exif: false,
      });

      if (!result.canceled && result.assets[0].uri) {
        const data = new FormData();

        console.log(result.assets[0].uri);
        const req = await fetch(result.assets[0].uri);

        const blob = await req.blob();

        data.append("avatar", blob);

        console.log(await pb.collection("users").update<IUser>(currentUser?.model?.id ?? "", data));
        console.log('Uploaded "successfully"');
      }
      setLoading(false);
    } catch (err) {
      setLoading(false);
      setError(true);
      console.error(err);
    }
  };
```

This happens with both react-native's FormData, and the FormData module I was using.

**Reply by EverlyScott** [reply]
I will try updating my pocketbase instance to the latest version and see if that fixes anything but I don't think that's the issue because it works on our website.

**Reply by EverlyScott** [reply]
> I will try updating my pocketbase instance to the latest version and see if that fixes anything but I don't think that's the issue because it works on our website.

Didn't work

**Reply by ganigeorgiev** [reply]
I'm not sure how to help you debug it. You can try wrapping the blob into a file, eg. `new File([blob], "test.png")` and see if that will make any difference.

If that doesn't help, please try to provide a minimal reproducible repo for example using https://snack.expo.dev/.

**Reply by EverlyScott** [reply]
Alright will do

**Reply by ganigeorgiev** [reply]
Optionally, you can also try starting PocketBase with `--debug` and see if there is any debug message for invalid file(s) printed in the console during the upload.

**Reply by EverlyScott** [reply]
I've only just gotten to this.
The only thing it logs in debug is this:
```sql
UPDATE `users` SET `avatar`=''
```

I'm currently building a snack to debug it further.

**Reply by EverlyScott** [reply]
Here is the snack link: [https://snack.expo.dev/@everlys/pocketbase-expo-avatar-upload-issues](https://snack.expo.dev/@everlys/pocketbase-expo-avatar-upload-issues)

**Reply by ganigeorgiev** [reply]
@ScribbleNerd I'm not able to reproduce the issue. Make sure that:

- your collection file field is named `avatar`
- the uploaded file size is less than or equal than the "Max size" specified in the field options
- the file mimetype is from the allowed mimetypes in the file field options (if any)

**Reply by ganigeorgiev** [reply]
@ScribbleNerd Actually I tested it again and it works with the web emulator but doesn't work on Android for some reason. It seems that there is something platform specific to the `FormData`/`fetch` React Native or Expo implementation and this would explain why the other user in https://github.com/pocketbase/pocketbase/discussions/1916 is using axios for sending plain old xhr requests instead.

**Comment by ganigeorgiev** [comment, marked as answer]
@ScribbleNerd I think I found the culprit. There are actually 2 issues:

1. React Native on Android and iOS seems to use a non-standard `FormData` implementation (_`set` and other similar methods for example are not available_) and in addition the `constructor.name` is mangled (_this could be Expo specific, but it returns the correct name with the web emulator_). To workaround this, I've pushed a minor [JS SDK fix in v0.12.1](https://github.com/pocketbase/js-sdk/commit/c5941bcac4ef977b43c424f3a9ddff814d0715b5#diff-b7961e132a9ada7fe5b4cdec69198fc7e4fe506e45b5a7fe29f399a01c45ff53).

2. While testing on Android, I wasn't able to fetch the blob with the uri, but  Android and iOS supports a special object like `FormData` value syntax for submitting local files:
```js
{
    uri: "...",
    type: "...",
    name: "...",
}
```

To summarize:

1. Update your JS SDK to v0.12.1 (_note that I had some cache issues with snack.expo and wasn't able to pull the latest package version for some reason and had to load the bundle file manually_)

2. When loading your FormData data use something like:
  ```js
  const data = new FormData();

  let imageUri = result.assets[0].uri;

  if (Platform.OS === 'web') {
    const req = await fetch(imageUri);
    const blob = await req.blob();
    data.append('avatar', blob);
  } else {
    // passing an object works only on Android and iOS
    data.append('avatar', {
      uri: imageUri,
      type: `image/*`, 
      name: imageUri.split('/').pop(),
    });
  }
  ```

**Reply by EverlyScott** [reply]
That works thank you so much!
